File: 1073-number-of-enclaves/1073-number-of-enclaves.java
import java.util.*;

class Solution {
    private int[] delRow = {-1, 0, 1, 0};
    private int[] delCol = {0, 1, 0, -1};

    private boolean isValid(int i, int j, int N, int M){
        if(i < 0 || i >= N) return false;
        if(j < 0 || j >= M) return false;

        return true;
    }

    private void BFS(int[][] grid, Queue<int[]> que, boolean[][] vis){
        int N = grid.length;
        int M = grid[0].length;

        while(!que.isEmpty()){
            int[] cell = que.poll();

            int row = cell[0];
            int col = cell[1];

            for(int i=0;i< 4;i++){
                int nRow = row + delRow[i];
                int nCol = col + delCol[i];

                if(isValid(nRow, nCol, N, M) &&
                   grid[nRow][nCol] == 1 
                   && vis[nRow][nCol] == false){
                    vis[nRow][nCol] = true;
                    que.add(new int[]{nRow, nCol});
                   }
            }
        }
    }
    public int numEnclaves(int[][] grid) {
        int N = grid.length;
        int M = grid[0].length;

        Queue<int[]> que = new LinkedList<>();

        boolean[][] vis = new boolean[N][M];

        for(int i=0;i < N; i++){
            for(int j=0; j< M; j++){

                if((i==0 || i == N -1 ||
                    j==0 || j == M -1) &&
                    grid[i][j] == 1){
                        vis[i][j] = true;
                        que.add(new int[]{i, j});
                    }
            }
        }
        BFS(grid, que, vis);

        int count = 0;

        for(int i=0;i<N;i++){
            for(int j=0;j<M;j++){
                if(grid[i][j] == 1 && vis[i][j] == false){
                    count++;
                }
            }
        }
        return count;
    }
}

File: 1073-number-of-enclaves/1073-number-of-enclaves.py
from collections import deque

class Solution:
    def __init__(self):
        self.delRow = [-1, 0, 1, 0]
        self.delCol = [0, 1, 0, -1]

    def isValid(self, i, j, N, M):
        if i < 0 or i >= N:
            return False 
        if j < 0 or j >= M:
            return False 
        
        return True 

    def BFS(self, grid, que, vis):
        N = len(grid)
        M = len(grid[0])

        while que:
            row, col = que.popleft()

            for i in range(4):
                nRow = row + self.delRow[i]
                nCol = col + self.delCol[i]

                if(
                    self.isValid(nRow, nCol, N, M) and
                    grid[nRow][nCol] == 1 and
                    vis[nRow][nCol] == False
                ):
                    vis[nRow][nCol] = True 
                    que.append((nRow, nCol))
            
    def numEnclaves(self, grid: List[List[int]]) -> int:
        N = len(grid)
        M = len(grid[0])

        que = deque()

        vis = [[False] * M for _ in range(N)]

        for i in range(N):
            for j in range(M):

                if(
                    (i == 0 or i == N - 1 or j == 0 or j == M - 1) and
                    grid[i][j] == 1
                ):
                    vis[i][j] = True
                    que.append((i, j))
        self.BFS(grid, que, vis)
        count = 0

        for i in range(N):
            for j in range(M):

                if(
                    grid[i][j] == 1 and
                    vis[i][j] == False
                ):
                    count += 1
        return count